### DISCUSSION TITLE
Help with JazzOAuthClient

### DISCUSSION THREAD

**Discussion by forsmann** [OSLC/oslc4net#146] [Q&A]
Hello,
I'm trying to use JazzRootServicesHelper.  I'm getting this error: Unable to access the Jazz rootservices document at:

However, when i put the url in the browser i get the rdf file.  It's an https site.  I put this in the code:

                ServicePointManager.ServerCertificateValidationCallback +=
                    (sender, cert, chain, sslPolicyErrors) => true;

Any ideas?

**Comment by berezovskyi** [comment]
Hi @forsmann,

The message you specified should be printed on `RootServicesException re` exception. What message does it come with and what is its nested exception?

You disable cert validation, does it mean your server has a self-signed TLS cert? Does your code work if you fetch rootservices from `https://oslc.itm.kth.se/ccm`? Please note that the library appends `/rootservices` on its own. Also, the way you disable cert validation is a bit different from passing a lambda to the oslc4net client, did you make sure your delegate gets called in the process?

I am assuming you are on .NET Framework (4.8?) due to Jazz project not having been ported to .NET 6+ yet. What version of the `OSLC4Net.Client` are you using? [0.3.0-alpha](https://www.nuget.org/packages/OSLC4Net.Client/0.3.0-alpha)?

On a side note, only https://github.com/OSLC/oslc4net/blob/main/OSLC4Net_NETFramework/OSLC4Net.Client.Jazz/Oslc/Jazz/JazzOAuthClient.cs requires .NET Framework. I am thinking to split `OSLC4Net.Client.Jazz` into:

- `OSLC4Net.Client.Jazz` that will be .NET 6+ compatible and have all the code except for https://github.com/OSLC/oslc4net/blob/main/OSLC4Net_NETFramework/OSLC4Net.Client.Jazz/Oslc/Jazz/JazzOAuthClient.cs, and
- `OSLC4Net.Client.Jazz.OAuth10a` that will only contain the https://github.com/OSLC/oslc4net/blob/main/OSLC4Net_NETFramework/OSLC4Net.Client.Jazz/Oslc/Jazz/JazzOAuthClient.cs class. It will remain to be .NET Framework dependent because DotNetOpenAuth only targets .NET Framework and I am not planning to fork&port it to .NET 6+ because it's using an obsolete OAuth1.0a mechanism. 

This should make Jazz client usable in .NET 6+, esp. if you rely on Jazz Forms auth (OIDC support is not yet part of oslc4net).

**Reply by forsmann** [reply]
Thanks for the quick reply.  I would like to run this on dotnet for Linux.  The readme states: 
"If you do not have a .NET development environment, start by downloading VS Code [C# Dev Kit](https://marketplace.visualstudio.com/items?itemName=ms-dotnettools.csdevkit). Make sure to install .NET 8 SDK for development. Libraries target NETStandard 2.0 and should run on .NET Framework 4.8 or .NET 6+ (recommended)."

Dotnet 4.X doesn't exist on Linux.  Will it work with 3.1?  I want to run the Jazz client with OAuth.

**Reply by forsmann** [reply]
Will it work in Mono?

**Reply by berezovskyi** [reply]
All the code under https://github.com/OSLC/oslc4net/tree/main/OSLC4Net_SDK will run on .NET 6/7/8 on Linux (I am developing myself on macOS and occasionally on openSUSE Tumbleweed and we have CI builds matrix for {net6,net7,net8}x{windows,ubuntu}, e.g. https://github.com/OSLC/oslc4net/actions/runs/8664390156 ). 

But the code under https://github.com/OSLC/oslc4net/tree/main/OSLC4Net_NETFramework has not yet been upgraded (reasons for Jazz project are listed in my original response) and require .NET Framework (or Mono). If I recall correctly, at least on macOS I failed to get the original project targeting .net framework 4.7 to run under mono and had to give up to do the migration on Windows. But maybe you have better luck (and chops to get the project working under mono).

The big question is whether you need the https://github.com/OSLC/oslc4net/blob/main/OSLC4Net_NETFramework/OSLC4Net.Client.Jazz/Oslc/Jazz/JazzOAuthClient.cs class and OAuth1.0a support? If not, better to do the migration I described above and to use .net6/8 on linux directly without mono.

But to be honest, I am surprised how you got that code running in the first place. Could you please post for the environment where you got the "Unable to access the Jazz rootservices document at" error:

- output of `uname -psrm` (sh) or `ver` (win cmd)
- output of `dotnet --info` - or mono
- versions of oslc4net nuget packages you are using or if you are using oslc4net from source - output from `git describe --tags --always --dirty --long` 
- full error message including the stack trace, only anonymizing hostnames, ips, usernames
- which sample you are running if it's one of https://github.com/OSLC/oslc4net/tree/main/OSLC4Net_NETFramework/OSLC4Net.Client.Samples . Or are you running the sample from https://github.com/OSLC/oslc4net/tree/main?tab=readme-ov-file#a-simple-oslc-client ?

**Reply by berezovskyi** [reply]
Ah now I saw you want OAuth (1.0a, not 2.0, right?). I think you can try with mono, the project does not involve using any specific Windows/GDI/WPF APIs.

**Reply by berezovskyi** [reply]
And to be clear, [OSLC4Net.Client.Jazz](https://github.com/OSLC/oslc4net/tree/main/OSLC4Net_NETFramework/OSLC4Net.Client.Jazz) will not work under .NET (Core) 3.1 because .NET 3.1 is _newer_ than .NET **Framework** 4.8. More on this under https://en.wikipedia.org/wiki/.NET#History. Everything under https://github.com/OSLC/oslc4net/tree/main/OSLC4Net_SDK except for test projects targets netstandard2.0 and will run on any reasonably modern dotnet, whether .net framework 4.8, or mono, or .net core 3.1 (don't use it in prod, it does not get security updates), or (best) under dotnet6 or dotnet8.

**Reply by forsmann** [reply]
I think I was headed down the wrong path. I would like to use .net core if possible. I see that OSLC4net_SDK should work.  I just didn't see an example of how to use OAuth and get root services using this.  Maybe you can help.

**Reply by berezovskyi** [reply]
A few points:

- There is no need to have OAuth to get OSLC root services. The root services document includes bootstrap information, possibly for oauth endpoints, and thus shall be accessible without auth.
- To get the root services document, you should use https://github.com/OSLC/oslc4net/blob/main/OSLC4Net_NETFramework/OSLC4Net.Client.Jazz/Oslc/Jazz/JazzRootServicesHelper.cs For now, that class is located in the project that only targets legacy .NET Framework / Mono. We would need to split the OAuth out of that project to make `JazzRootServicesHelper` directly usable in .net core.
- There is no way to support OAuth**1.0a** oslc4net code in .net core because that code depends on https://www.nuget.org/packages/DotNetOpenAuth.OAuth.Consumer/4.3.4.13329#supportedframeworks-body-tab and it does not support .net core. Possible solutions (from simplest to ):
	- use Jazz Forms auth
	- use OAuth 2 / OIDC (using RestSharp or only via ASP.NET Core as shown in https://github.com/magnusstuhr/httpclient-seamless-openid-oauth2/tree/main )
	- rewrite the OAuth1.0a code in oslc4net to use RestSharp that provides OAuth 1 and 2 clients: https://restsharp.dev/authenticators.html#oauth1 – I will be happy to review and merge the PR if you are ready to help out with this
	- fork DotNetOpenAuth.OAuth.Consumer and port that code to netstandard2.0 or net6.0+ (questionable from security POV, that library is dead for at least ~7~ 10 years).

**Reply by berezovskyi** [reply]
What OSLC application are you integrating with? Is it IBM Jazz?

**Reply by forsmann** [reply]
IBM DOORS.  I just got the Dot Net framework version building in Mono.  Going to try it.
